server/core/runtime/run_step_runner.go

```diff
@@ -66,13 +73,45 @@ func (r *RunStepRunner) Run(ctx command.ProjectContext, command string, path str
 		finalEnvVars = append(finalEnvVars, fmt.Sprintf("%s=%s", key, val))
 	}
 	cmd.Env = finalEnvVars
-	out, err := cmd.CombinedOutput()
+	stdout, err := cmd.StdoutPipe()
+	if err != nil {
+		err = fmt.Errorf("%s: unable to create stdout buffer", err)
+		ctx.Log.Debug("error: %s", err)
+		return "", err
+	}
+	stderr, err := cmd.StderrPipe()
+	if err != nil {
+		err = fmt.Errorf("%s: unable to create stderr buffer", err)
+		ctx.Log.Debug("error: %s", err)
+		return "", err
+	}
+	cmd.Start()
+
+	var output bytes.Buffer
+	var mutex sync.Mutex
+
+	go r.streamOutput(ctx, stdout, &output, &mutex)
+	go r.streamOutput(ctx, stderr, &output, &mutex)
+
+	err = cmd.Wait()
 
 	if err != nil {
-		err = fmt.Errorf("%s: running %q in %q: \n%s", err, command, path, out)
+		err = fmt.Errorf("%s: running %q in %q: \n%s", err, command, path, output.String())
 		ctx.Log.Debug("error: %s", err)
 		return "", err
 	}
 	ctx.Log.Info("successfully ran %q in %q", command, path)
-	return string(out), nil
+	return ansi.Strip(output.String()), nil
+}
+
+func (r RunStepRunner) streamOutput(ctx command.ProjectContext, reader io.Reader, buffer *bytes.Buffer, mutex *sync.Mutex) {
+	scanner := bufio.NewScanner(reader)
+	for scanner.Scan() {
+		line := scanner.Text()
+		r.ProjectCmdOutputHandler.Send(ctx, line, false)
```

Comment:
I noticed that the `OperationComplete` bool was added in #2051 but don't see it ever being set to `true` in `terraform_client.go`. I'm guessing this means we might be leaking goroutines -- should this emit a final empty line with `true`?

Reply:
nevermind, i see this is handled in `project_command_runner.go`